### PR TITLE
Yii2-notifications improvements

### PR DESCRIPTION
**Pull request by KarimaJarmouni**
Pull request

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "intelephense.environment.includePaths": ["/var/www/html/yetopen/basic/vendor"]
```
Comment (liviuk2) [inline .vscode/settings.json]: remove specific files of your editor from git, ad them in .gitignore

File: src/Module.php [hunk 2]
```diff
@@ -49,19 +49,25 @@ public function init()
      * @return bool If the sending was successful or not.
      */
     public function send($notification, array $channels = null){
+
```
Comment (liviuk2) [inline src/Module.php:51]: remove unnecessary empty lines

File: src/Notification.php [hunk 3]
```diff
@@ -19,6 +20,8 @@ class Notification extends \yii\base\BaseObject
 
     public $userId = 0;
 
+    public $typeCode = 'MM';
```
Comment (liviuk2) [inline src/Notification.php]: the default type should be "Generic" whit code 'GN'

File: src/channels/ScreenChannel.php [hunk 4]
```diff
@@ -13,6 +13,7 @@ class ScreenChannel extends Channel
      */
     public function sendNotification($notification)
     {
+        $notification->getTitle();
```
Comment (liviuk2) [inline src/channels/ScreenChannel.php]: what is this change for?

File: src/model/NotificationType.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+ * @property integer|null $priority
+ */
+class NotificationType extends \yii\db\ActiveRecord
+{
+    /**
+     * {@inheritdoc}
+     */
+    public static function tableName()
+    {
+        return 'notifications_type';
```
Comment (liviuk2) [inline src/model/NotificationType.php]: Properly quote table name: https://www.yiiframework.com/doc/guide/2.0/en/db-dao#quoting-table-and-column-names

File: src/views/notification-type/index.php [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                'value'=> function ($model, $key, $index, $column){
+                    return $model->getPrio($model->priority);
+                },
+                'vAlign' => 'middle',
+                'hAlign' => 'center',
+            ],
+            [
+                'class' => ActionColumn::className(),
+                'template' => '{update}',
+                'urlCreator' => function ($action, webzop\notifications\model\NotificationType $model, $key, $index, $column) {
```
Comment (liviuk2) [inline src/views/notification-type/index.php:87]: why is this needed?

File: src/views/notification-type/index.php [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+use yii\grid\ActionColumn;
+use \kartik\grid\GridView;
+use webzop\notifications\model\NotificationType;
+
+
+/* @var $this yii\web\View */
+/* @var $searchModel webzop\notifications\model\NotificationTypeSearch */
+/* @var $dataProvider yii\data\ActiveDataProvider */
+
+$this->title = 'Notification Types';
```
Comment (liviuk2) [inline src/views/notification-type/index.php]: This strings are not translatable, please wrap them with `Yii::t` method. This must be done for all the strings that must be translated

File: src/views/notification-type/index.php [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                'vAlign'=>'middle',
+                'format'=>'raw',
+                'width'=>'200px',
+                'noWrap'=>true,
+                'vAlign' => 'middle',
+                'hAlign' => 'center',
+            ],
+            'priority' => [
+                'attribute' => 'priority',
+                'filter' => [0=>'Low', 1=>'Medium', 2=>'High'],
```
Comment (liviuk2) [inline src/views/notification-type/index.php]: Is not a good practice to define the same dictionary in more than one place, please use: https://code.ufficyo.com/YetOpen/yii2-extensions/yii2-helpers/-/blob/master/dictionaries/Dictionary.php

See: https://code.ufficyo.com/YetOpen/yii2-extensions/yii2-helpers/-/blob/master/interfaces/DictionaryInterface.php

File: src/views/notification/index.php [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+                'width'=>'170px',
+            ],
+            'message' => [
+                'attribute' => 'message',
+                'hAlign' => 'center',
+                'width'=>'200px',
+            ],
+            'created_at' => [
+                'attribute' => 'created_at',
+                'label' => 'Datetime',
```
Comment (liviuk2) [inline src/views/notification/index.php]: Better to use the label defined on the attribute in the model

File: src/views/notification/index.php [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+                    ]
+                ]),
+                'vAlign' => 'middle',
+                'hAlign' => 'center',
+            ],
+
+            'timeago' => [
+                'attribute' => 'timeAgo',
+                'value' => function ($model, $key, $index, $column){
+                    return $model->timeAgo;
```
Comment (liviuk2) [inline src/views/notification/index.php]: this is not needed, by specifying the attribute 'timeAgo' it automatically returns ` $model->timeAgo`

File: src/widgets/Notifications.php [hunk 11]
```diff
@@ -130,7 +130,7 @@ public static function getCountUnseen(){
         $userId = Yii::$app->getUser()->getId();
         $count = NotificationModel::find()
             ->andWhere(['or', 'user_id = 0', 'user_id = :user_id'], [':user_id' => $userId])
-            ->andWhere(['<=', 'sent_at', date('Y-m-d H:i:s')])
+            ->andWhere(['<=', 'send_at', date('Y-m-d H:i:s')])
```
Comment (liviuk2) [inline src/widgets/Notifications.php:133]: why this change? if you count the unseen notifications you want to count only the onces already sent so it's ok to use 'sent_at' attribute

File: src/views/notification/index.php [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+                ],
+                'vAlign' => 'middle',
+                'hAlign' => 'center',
+                
+            ],
+            'managed' => [
+                'attribute'=>'managed', 
+                'class' => 'kartik\grid\BooleanColumn',
+                'class' => 'kartik\grid\EditableColumn',
+                'filter' => [0 => 'Unmanaged', 1 => 'Managed'],
```
Comment (liviuk2) [inline src/views/notification/index.php]: apply the dictionary logic

File: src/views/notification/index.php [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+                
+                'editableOptions'=> function($model, $key, $index, $column){
+                
+                    if(($model->notificationsType->check_management) == 0){
+                        return [
+                            'editableButtonOptions' => ['disabled' => true,],
+                            'asPopover' => true,
+                            'format' => Editable::FORMAT_BUTTON,
+                            'inputType' => Editable::INPUT_DROPDOWN_LIST,
+                            'data' => [0 => 'Unmanaged', 1 => 'Managed'],
```
Comment (liviuk2) [inline src/views/notification/index.php]: apply the dictionary logic

File: src/views/notification/index.php [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+                'filter' => [0 => 'Unread', 1 => 'Read'],
+                'filterInputOptions' => [
+                    'class' => 'form-control',         
+                    'prompt' => 'All'
+                ],
+                'editableOptions'=> [
+                    'asPopover' => true,
+                    'format' => Editable::FORMAT_BUTTON,
+                    'inputType' => Editable::INPUT_DROPDOWN_LIST,
+                    'data' => [0 => 'Unread', 1 => 'Read'],
```
Comment (liviuk2) [inline src/views/notification/index.php]: apply the dictionary logic

File: src/views/notification/index.php [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+                }, 
+                'vAlign' => 'middle',
+                'hAlign' => 'center',
+                'width' => '200'
+            ],
+            'read' =>  [
+                'attribute'=>'read',
+                'class' => 'kartik\grid\BooleanColumn',
+                'class' => 'kartik\grid\EditableColumn',
+                'filter' => [0 => 'Unread', 1 => 'Read'],
```
Comment (liviuk2) [inline src/views/notification/index.php]: apply the dictionary logic

File: src/views/notification/index.php [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+/* @var $this yii\web\View */
+/* @var $searchModel webzop\notifications\model\NotificationSearch */
+/* @var $dataProvider yii\data\ActiveDataProvider */
+
+
+$this->title = 'Notifications';
+$this->params['breadcrumbs'][] = $this->title;
+?>
+<div class="notifications-index">
+    <script defer src="https://use.fontawesome.com/releases/v5.3.1/js/all.js" crossorigin="anonymous"></script>
```
Comment (liviuk2) [inline src/views/notification/index.php]: this should be a requirement in your composer, not specified directly in the view. Use: https://github.com/kartik-v/yii2-icons

File: src/views/notification/index.php [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+                        ];
+                    }
+
+                    if(($model->notificationsType->check_management) == 1){
+                        return [
+                            'editableButtonOptions' => ['disabled' => false,],
+                            'asPopover' => true,
+                            'format' => Editable::FORMAT_BUTTON,
+                            'inputType' => Editable::INPUT_DROPDOWN_LIST,
+                            'data' => [0 => 'Unmanaged', 1 => 'Managed'],
```
Comment (liviuk2) [inline src/views/notification/index.php]: apply the dictionary logic

File: .vscode/settings.json (hunk 1 above)
Comment (liviuk2) [inline .vscode/settings.json]: do not commit IDE config files, remove it from git and add the file in .gitignore

**Review by liviuk2 — CHANGES_REQUESTED** [review]
- missing documentation on how to use new features
- add new version and changelog

File: src/model/NotificationSearch.php [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+            ->andFilterWhere(['like', 'message', $this->message])
+            ->andFilterWhere(['like', 'content', $this->content])
+            ->andFilterWhere(['like', 'attachments', $this->attachments])
+            ->andFilterWhere(['like', 'language', $this->language])
+            ->andFilterWhere(['like', 'route', $this->route]);
+
+
+        //filter datetime range from string to integer
+        if(isset ($this->created_at_filter) && $this->created_at_filter != '' && strpos($this->created_at_filter, ' - ') !== false){
+            $date_explode = explode(" - ", $this->created_at_filter);
```
Comment (mp1509) [inline src/model/NotificationSearch.php:109]: It'd be better to use [DateTime::createFromFormat()](https://www.php.net/manual/en/datetime.createfromformat.php), see [our knowledge base guide](https://kb.ufficyo.com/en/php/yii2/utilities-and-tips#add-daterangefilter)

File: src/views/notification/index.php [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+                'label' => 'Datetime',
+                'format' => 'datetime',
+                'filter' => DateRangePicker::widget([
+                    'model' => $searchModel,
+                    'attribute' => 'created_at_filter',
+                    'convertFormat' => true,
+                    'pluginOptions' => [
+                        'timePicker'=>true, 
+                        'timePickerIncrement'=>10,
+                        'locale' => ['format' => 'd-m-Y g:i A']
```
Comment (mp1509) [inline src/views/notification/index.php]: The format should be a constant of the search model so that it can be reused there (see my previous comment). Also, `d-m-Y g:i A` is a bit strange as a format, usually it's used `d/m/Y H:i` (01/01/2000 00:00). About this, @liviuk2 it's an open point since it should be used the locale's preferred format, for now I'd set the Italian standard

**Review by mp1509 — CHANGES_REQUESTED** [review]
